File: app/jql_builder.py
from __future__ import annotations

import json
from typing import Any

import httpx

from app.config import Settings
from app.http_util import raise_for_status


def _jql_field_reference(field: dict[str, Any]) -> str | None:
    """Stable JQL reference: cf[n] for customfield_n; else id or quoted name."""
    fid = field.get("id")
    if isinstance(fid, str) and fid.startswith("customfield_"):
        suffix = fid[len("customfield_") :]
        if suffix.isdigit():
            return f"cf[{suffix}]"
    if isinstance(fid, str) and fid and " " not in fid:
        return fid
    name = field.get("name")
    if name:
        return json.dumps(str(name))
    return None


def _field_clauses_for_current_user(
    field: dict[str, Any], api_major: int
) -> list[str]:
    """Build JQL fragments for fields where currentUser() is meaningful."""
    if not field.get("searchable", False):
        return []

    fid = field.get("id")
    # Data Center REST often marks fields "searchable" that JQL still rejects (sort-only
    # cf[], internal ids like archivedby). Cloud-style OR-of-all-user-fields is unsafe.
    if api_major == 2:
        return []

    quoted = _jql_field_reference(field)
    if not quoted:
        return []

    schema = field.get("schema") or {}
    ftype = schema.get("type")
    items = (schema.get("items") or "").lower()

    if ftype == "user":
        return [f"{quoted} = currentUser()"]

    if ftype == "array" and items in {"user", "sd-user"}:
        # multi-user picker
        return [f"{quoted} in (currentUser())"]

    return []


def build_jql_fragments(
    http: httpx.Client, settings: Settings, api_major: int
) -> list[str]:
    url = f"{settings.jira_site}/rest/api/{api_major}/field"
    resp = http.get(url)
    raise_for_status(resp, f"Jira GET /rest/api/{api_major}/field")

    fields: list[dict[str, Any]] = resp.json()
    clauses: list[str] = [
        "assignee = currentUser()",
        "reporter = currentUser()",
    ]

    # watcher / participant: Jira Cloud / JSD; often missing on Server/Data Center
    if api_major == 3:
        clauses.append("watcher = currentUser()")
        clauses.append("participant in (currentUser())")

    seen: set[str] = set(clauses)
    for field in fields:
        for c in _field_clauses_for_current_user(field, api_major):
            if c not in seen:
                seen.add(c)
                clauses.append(c)

    return clauses


def merge_jql_or(fragments: list[str]) -> str:
    parts = [f"({f})" for f in fragments if f.strip()]
    return " OR ".join(parts)


def _chunk_fragments(fragments: list[str], max_chars: int = 6500) -> list[list[str]]:
    """Split OR clauses so each JQL substring stays under the approximate size limit."""
    groups: list[list[str]] = []
    buf: list[str] = []
    size = 0
    for frag in fragments:
        frag = frag.strip()
        if not frag:
            continue
        piece = f"({frag})"
        add = len(piece) + (4 if buf else 0)  # " OR "
        if buf and size + add > max_chars:
            groups.append(buf)
            buf = [frag]
            size = len(piece)
        else:
            buf.append(frag)
            size += add
    if buf:
        groups.append(buf)
    return groups


def jira_jql_batches(
    settings: Settings, http: httpx.Client, api_major: int
) -> list[str]:
    """One or more JQL queries; the sync layer deduplicates issue keys across batches."""
    if settings.use_raw_jql_only and settings.raw_jql:
        return [settings.raw_jql]

    fragments = build_jql_fragments(http, settings, api_major)
    merged = merge_jql_or(fragments)
    if settings.extra_jql:
        merged = f"({merged}) OR ({settings.extra_jql})"

    if len(merged) <= 7200:
        return [merged]

    batches = [merge_jql_or(g) for g in _chunk_fragments(fragments) if g]
    if settings.extra_jql:
        batches.append(settings.extra_jql)
    return batches

File: node/src/jira/jqlBuilder.js
import { raiseForStatus, baseFetchInit } from "../http.js";

/**
 * @param {Record<string, unknown>} field
 * @returns {string | null}
 */
function jqlFieldReference(field) {
  const fid = field.id;
  if (typeof fid === "string" && fid.startsWith("customfield_")) {
    const suffix = fid.slice("customfield_".length);
    if (/^\d+$/.test(suffix)) return `cf[${suffix}]`;
  }
  if (typeof fid === "string" && fid && !fid.includes(" ")) return fid;
  const name = field.name;
  if (name) return JSON.stringify(String(name));
  return null;
}

/**
 * @param {Record<string, unknown>} field
 * @param {number} apiMajor
 * @returns {string[]}
 */
function fieldClausesForCurrentUser(field, apiMajor) {
  if (!field.searchable) return [];
  if (apiMajor === 2) return [];

  const quoted = jqlFieldReference(field);
  if (!quoted) return [];

  const schema = /** @type {Record<string, string>} */ (field.schema || {});
  const ftype = schema.type;
  const items = (schema.items || "").toLowerCase();

  if (ftype === "user") return [`${quoted} = currentUser()`];
  if (ftype === "array" && (items === "user" || items === "sd-user")) {
    return [`${quoted} in (currentUser())`];
  }
  return [];
}

/**
 * @param {Record<string, unknown>} settings
 * @returns {Promise<2 | 3>}
 */
export async function resolveJiraApiMajor(settings) {
  const forced = settings.jiraRestApiVersion;
  if (forced === 2 || forced === 3) return forced;

  const base = String(settings.jiraSite).replace(/\/+$/, "");
  const r3 = await fetch(`${base}/rest/api/3/field`, baseFetchInit(settings));
  if (r3.status === 200) return 3;
  if (r3.status === 404) {
    const r2 = await fetch(`${base}/rest/api/2/field`, baseFetchInit(settings));
    if (r2.status === 200) return 2;
    if (r2.status === 404) {
      console.error(
        "Jira REST returned 404 for both /rest/api/3/field and /rest/api/2/field. " +
          "If ATLASSIAN_SITE is your Confluence URL, set JIRA_SITE to the Jira base URL " +
          "(browser address bar on the Jira dashboard, no trailing slash)."
      );
      process.exit(1);
    }
    await raiseForStatus(r2, "Jira GET /rest/api/2/field (v3 unavailable)");
    return 2;
  }
  await raiseForStatus(r3, "Jira GET /rest/api/3/field");
  return 3;
}

/**
 * @param {Record<string, unknown>} settings
 * @param {2 | 3} apiMajor
 * @returns {Promise<string[]>}
 */
export async function buildJqlFragments(settings, apiMajor) {
  const base = String(settings.jiraSite).replace(/\/+$/, "");
  const url = `${base}/rest/api/${apiMajor}/field`;
  const res = await fetch(url, baseFetchInit(settings));
  await raiseForStatus(res, `Jira GET /rest/api/${apiMajor}/field`);
  /** @type {Record<string, unknown>[]} */
  const fields = await res.json();

  const clauses = [
    "assignee = currentUser()",
    "reporter = currentUser()",
  ];
  if (apiMajor === 3) {
    clauses.push("watcher = currentUser()");
    clauses.push("participant in (currentUser())");
  }

  const seen = new Set(clauses);
  for (const field of fields) {
    for (const c of fieldClausesForCurrentUser(field, apiMajor)) {
      if (!seen.has(c)) {
        seen.add(c);
        clauses.push(c);
      }
    }
  }
  return clauses;
}

/** @param {string[]} fragments */
export function mergeJqlOr(fragments) {
  const parts = fragments.filter((f) => f.trim()).map((f) => `(${f})`);
  return parts.join(" OR ");
}

/** @param {string[]} fragments @param {number} [maxChars] */
export function chunkFragments(fragments, maxChars = 6500) {
  /** @type {string[][]} */
  const groups = [];
  let buf = [];
  let size = 0;
  for (let frag of fragments) {
    frag = frag.trim();
    if (!frag) continue;
    const piece = `(${frag})`;
    const add = piece.length + (buf.length ? 4 : 0);
    if (buf.length && size + add > maxChars) {
      groups.push(buf);
      buf = [frag];
      size = piece.length;
    } else {
      buf.push(frag);
      size += add;
    }
  }
  if (buf.length) groups.push(buf);
  return groups;
}

/**
 * @param {Record<string, unknown>} settings
 * @param {2 | 3} apiMajor
 * @returns {Promise<string[]>}
 */
export async function jiraJqlBatches(settings, apiMajor) {
  if (settings.useRawJqlOnly && settings.rawJql) {
    return [settings.rawJql];
  }

  const fragments = await buildJqlFragments(settings, apiMajor);
  let merged = mergeJqlOr(fragments);
  if (settings.extraJql) {
    merged = `(${merged}) OR (${settings.extraJql})`;
  }
  if (merged.length <= 7200) return [merged];

  const batches = chunkFragments(fragments).map((g) => mergeJqlOr(g));
  if (settings.extraJql) batches.push(settings.extraJql);
  return batches;
}
